Compare against experiment as well

chriskrycho opened the pull request:
- add a comparison for local fork of @glimmer/compiler
- avoid holding onto templates so we can push on actual memory behavior using the same basic infrastructure: load them for each compiler pass, and 'forget' them by consuming them from the array of template strings while doing compilation

chriskrycho committed:
Compare against experiment as well

- add a comparison for local fork of @glimmer/compiler
- avoid holding onto templates so we can push on actual memory
  behavior using the same basic infrastructure: load them for
  each compiler pass, and 'forget' them by consuming them from
  the array of template strings while doing compilation

brendenpalmer commented on package.json:
```diff
@@ -15,8 +15,9 @@
-    "glimmer-compiler-ember-source-3-28": "npm:@glimmer/compiler@0.80.0",
     "glimmer-compiler-ember-source-3-24": "npm:@glimmer/compiler@0.65.3",
+    "glimmer-compiler-ember-source-3-28": "npm:@glimmer/compiler@0.80.0",
+    "glimmer-compiler-experiment": "link:../../glimmerjs/glimmer-vm/dist/@glimmer/compiler",
```
can you add this to the repo as a workspace?

chriskrycho replied:
Ah, that's a good thought. Will do!

chriskrycho replied:
Pushed it as a submodule, which should make it easier to work with over time. 🤞🏼

chriskrycho committed:
Add Glimmer VM as a submodule

rwjblue commented on bench/compare-template-compile-time.js:
```diff
@@ -34,13 +35,20 @@ function getCompilers() {
+          require(`${emberPackage}/dist/ember-template-compiler`).compileOptions;
+      } catch {
+        buildCompileOptions =
+          require(`ember-source-3-28/dist/ember-template-compiler`).compileOptions;
```
Mind making a "well named constant" for this? `const FallBackEmberSource`? Some better name?

(the idea being that it explains why this version is "special" / "magical" at the moment)

chriskrycho replied:
Yes, I like that.

chriskrycho replied:
Done!

chriskrycho committed:
Don't try to make the submodule a workspace for now

There's something not working here and it doesn't matter enough to dig
into it at this point in time for this little repro repo.

chriskrycho committed:
Give the fallback version a decent name

chriskrycho committed:
Run the compilation 5x for each compiler and average the results

chriskrycho committed:
Update workspace including debug setup

chriskrycho committed:
Update glimmer-vm submodule

chriskrycho committed:
Working CPU profiling and bench scripts with new submodules

chriskrycho committed:
Hide the mainline Glimmer VM repo in the workspace

There's no particular value in having it visible; it's just noise when
working here -- it's really only needed for the sake of easy comparison
with the experiment, in fact.

chriskrycho committed:
Time stamp CPU profiles when generating them (easy sorting!)

chriskrycho committed:
Better time stamping :joy:

chriskrycho committed:
Improve reporting (little Markdown table)

chriskrycho committed:
Run benchmark via `execa`

Previously, we saw issues where the results were ending up varying
wildly for some cases, likely because we were ending up with a dirty
heap, triggering GC, and so on. Now, run in a new process each time;
this produces much more consistent results. Curiously, it also shows
much higher run times, but the proportions are consistent with those
from previous analyses.

Combined with the changes in a previous commit to generate a table, and
adding in min and max values, the results currently look like this when
averaged over five separate runs for each source we care about:

|           variant            | average |  min  |  max  |
| ---------------------------- | ------- | ----- | ----- |
| ember-source@3.24.5          | 14381.6 | 13538 | 14790 |
| ember-source@3.25.4          | 44363.6 | 43375 | 45090 |
| ember-source@3.28.1          | 45368.2 | 45224 | 45484 |
| @glimmer/compiler@0.65.3     | 14493.8 | 14353 | 14600 |
| @glimmer/compiler@0.80.0     | 44395   | 44018 | 44679 |
| @glimmer/compiler@experiment | 20633.8 | 20523 | 20777 |
| @glimmer/compiler@latest     | 20751.4 | 20569 | 20864 |

chriskrycho committed:
Better time stamp; serialize the runs; skip some of them

chriskrycho committed:
Updated lockfile

chriskrycho committed:
Report median for runs
